File: go/scanner/module/subdomains/subdomains.go
package subdomains

import (
	"strings"

	"github.com/astrolib/govalidator"
	"github.com/bloom42/phaser/common/phaser"
	"github.com/bloom42/phaser/scanner/module"
	"github.com/jmoiron/sqlx"
	_ "github.com/lib/pq"
)

type Subdomains struct{}

func (Subdomains) Name() string {
	return "subdomains"
}

func (Subdomains) Description() string {
	return "Find subdomains for a given domain"
}

func (Subdomains) Author() string {
	return "Sylvain Kerkour <sylvain@kerkour.com>"
}

func (Subdomains) Version() string {
	return "0.1.0"
}

func reverseStr(s string) string {
	r := []rune(s)
	for i, j := 0, len(r)-1; i < len(r)/2; i, j = i+1, j-1 {
		r[i], r[j] = r[j], r[i]
	}
	return string(r)
}

type crtshDomain struct {
	Domain string `db:"domain"`
}

func (ports Subdomains) Run(scan *phaser.Scan, target *phaser.Target) (module.Result, []error) {
	uniqDomains := map[string]bool{}
	errs := []error{}
	var ret module.Result
	domains := []string{}

	crtshdb, err := sqlx.Connect("postgres", "host=crt.sh user=guest dbname=certwatch sslmode=disable")
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}

	subdomainsPattern := "%." + target.Host
	crtshDomains := []crtshDomain{}
	idx := strings.Index(subdomainsPattern, "%")
	idxrev := strings.Index(reverseStr(subdomainsPattern), "%")
	if idx != -1 && idx < idxrev {
		err = crtshdb.Select(&crtshDomains, `SELECT DISTINCT ci.NAME_VALUE as domain
			FROM certificate_identity ci
			WHERE reverse(lower(ci.NAME_VALUE)) LIKE reverse(lower($1))`, subdomainsPattern)
	} else {
		err = crtshdb.Select(&crtshDomains, `SELECT DISTINCT ci.NAME_VALUE as domain
			FROM certificate_identity ci
			WHERE lower(ci.NAME_VALUE) LIKE lower($1)`, subdomainsPattern)
	}
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}

	for _, crtshDomain := range crtshDomains {
		domain := strings.ToLower(strings.TrimSpace(crtshDomain.Domain))
		if govalidator.IsDNSName(domain) {
			uniqDomains[domain] = true
		}
	}

	for domain := range uniqDomains {
		domains = append(domains, domain)
	}
	ret = domains

	return ret, errs
}

File: src/scanner/modules/domain/subdomains.rs
use crate::{
    scanner::{
        module,
        findings,
        ReportV1,
        Target,
        TargetKind,
        ModuleName,
    },
    error::PhaserError,
};
use postgres::{Connection, TlsMode};

pub struct Subdomains{}

impl module::BaseModule for Subdomains {
    fn name(&self) -> ModuleName {
        return ModuleName::DomainSubdomains;
    }

    fn description(&self) -> String {
        return String::from("Find subdomains for a given domain");
    }

    fn author(&self) -> String {
        return String::from("Sylvain Kerkour <sylvain@kerkour.com>")
    }

    fn version(&self) -> String {
        return String::from("0.1.0");
    }
}

impl module::HostModule for Subdomains {
    fn run(&self, _: &ReportV1, target: &Target) -> Result<findings::Data, PhaserError> {
        let mut domains = vec!();

        if let TargetKind::Ip = target.kind {
            return Ok(findings::Data::None);
        };

        let conn = Connection::connect("postgres://guest@crt.sh:5432/certwatch", TlsMode::None)?;

        let subdomains_pattern = format!("%.{}", &target.host);

        let index = match subdomains_pattern.find('%') {
            Some(i) => i,
            None => 0,
        };
        let index_reverse = match subdomains_pattern.chars().rev().collect::<String>().find('%') {
            Some(i) => i,
            None => 0,
        };

        let query = if index < index_reverse {
            "SELECT DISTINCT ci.NAME_VALUE as domain
			FROM certificate_identity ci
			WHERE reverse(lower(ci.NAME_VALUE)) LIKE reverse(lower($1))"
        } else {
             "SELECT DISTINCT ci.NAME_VALUE as domain
            FROM certificate_identity ci
            WHERE lower(ci.NAME_VALUE) LIKE lower($1)"
        };

        let rows = conn.query(query, &[&subdomains_pattern])?;
        for row in &rows {
            domains.push(row.get(0));
        }

        return Ok(findings::Data::Domains(domains));
    }
}

#[cfg(test)]
mod tests {
    use crate::scanner::module::BaseModule;

    #[test]
    fn module_name() {
        let module = super::Subdomains{};
        assert_eq!("domain/subdomains", module.name().to_string());
    }
}
